aten/src/ATen/native/mps/OperationUtils.h

```diff
@@ -0,0 +1,196 @@
+MPSGraphTensor* trunc_tensor(MPSGraph* mpsGraph, MPSGraphTensor* inputTensor);
+MPSGraphTensorData *getMPSGraphTensorData(MPSGraph* mpsGraph,
+                                          MPSStream* mpsStream,
+                                          const Tensor& tensor);
+
+MPSGraph* make_mps_graph();
+void printTensorNDArray(const Tensor& t);
+
+MPSGraphTensor* mpsGraphUnrankedPlaceHolder(MPSGraph *mpsGraph, MPSDataType dataType);
+MPSGraphTensor* mpsGraphRankedPlaceHolder(MPSGraph *mpsGraph, MPSDataType dataType, MPSShape* mpsShape);
+MPSGraphTensor* mpsGraphRankedPlaceHolder(MPSGraph *mpsGraph, const Tensor& tensor);
+MPSGraphTensor* mpsGraphConstantFloatPlaceHolder(MPSGraph *mpsGraph, const double value, MPSShape* mpsShape);
+
+string get_mem_format_string(c10::MemoryFormat memory_format);
+
+using MPSCacheKey = int64_t;
+
+// derive this class to cache a graph and its inputs/ouputs
+// can be used to store any NSObject
+struct MPSCachedGraph
+{
+  MPSCachedGraph(NSObject *object) : _object([object retain]) {}
+  virtual ~MPSCachedGraph() {
+   [_object release];
+   _object = nullptr;
+  }
+  MPSGraph *graph() const { return (MPSGraph *)_object; }
+  NSObject *object() const { return _object; }
+private:
+  NSObject *_object = nullptr;
+};
+
+// Cache holding various keys mapped to graphs
+
+struct MPSGraphCache
+{
+  typedef MPSCachedGraph * (^CreateCachedGraphBlock)();
+
+  struct CacheEntry {
+    CacheEntry(std::string key, MPSCachedGraph *cachedGraph) : cachedGraph_(cachedGraph), key_(key) {}
+    MPSCachedGraph* cachedGraph_ = nullptr;
+    std::string key_ = nullptr;
+  };
+
+ public:
+
+  static MPSGraphCache* getInstance() {
+    if(_instance_cache == nullptr) {
+      _instance_cache = new MPSGraphCache();
+    }
+    return _instance_cache;
+  }
+
+  ~MPSGraphCache() {
+    dispatch_release(serialQueue_);
+
+    for (auto i : cache_) {
+      delete i.second.cachedGraph_;
+    }
+  }
+
+  // Disallow the copy constructor and operator= functions
+  MPSGraphCache(const MPSGraphCache&) = delete;
+  void operator=(const MPSGraphCache&) = delete;
+
+  MPSCachedGraph* CreateCachedGraph(const std::string& key, CreateCachedGraphBlock createCacheBlock) {
+
+    __block MPSCachedGraph * result = nil;
+
+    MPSCacheKey hash = std::hash<std::string>{}(key);
+
+    dispatch_sync(serialQueue_, ^() {
+
+      // verify the cached entry doesn't already exist
+      if (cache_.count(hash) != 0) {
+        auto& entry = cache_.at(hash);
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(key == entry.key_, "Key collision in the MPS cached graph!\n");
+        result = entry.cachedGraph_;
+      }
+      else {
+        result = createCacheBlock();
+        CacheEntry entry(key, result);
+        cache_.emplace(hash, entry);
+      }
+    });
+    return result;
+  }
+
+  MPSCachedGraph* LookUp(const std::string& key) const {
+
+    __block MPSCachedGraph* result = nullptr;
+
+    MPSCacheKey hash = std::hash<std::string>{}(key);
```

Comment:
This looks a bit questionable. If you're going to do a hash map from string to cached graph, why not just store the strings in the hash map directly; that way you don't have to manually test if the key matches, the hash map implementation will take care of it for you (in fact, it won't even complain about a key collision, and it will probe or whatever the implementation does.)

I'm also wondering a bit if you really want to do the cache on strings. Seems questionable.

Reply:
I do agree that we can improve the Caching strategy. But currently this design is deeply embedded into all the operations. I would add a TODO here and we can follow up to improve this strategy in future PR.